### PR TITLE
Fix issues with Texture2DAtlas cache system

### PR DESCRIPTION
**Pull request by FrancescoC-unity**
**TL;DR**: The atlas cache was marking everything as dirty every frame, causing every texture to be updated every frame. I assume this had a significant perf impact on projects with many cookies. 
No specific bug associated with this, I found the issues while investigating another bug fix. 

QA: You can skip to the bottom section for you. 

 Sorry for the wall of text, but since is a delicate change, I prefer describing too much than too little. 
The actual code changes are not much.  :P 

**The long version:**

So breaking the NeedsUpdate() function checks 

- The render texture case: was always returning true, this is fixed already by @JulienIgnace-Unity  in https://github.com/Unity-Technologies/Graphics/pull/1527 

- This check 

`            else if (m_TextureHashes.TryGetValue(key, out int hash) && hash != textureHash)`

Would beforehand always return true, the reason is because the hash was stored as uint in the dictionary, however textureHash was computed from int and cast to a uint while the value out of the dictionary was queried as int. As a result we were doing an int vs uint comparison  after cast of an hash which is likely gonna be very big numbers and quite likely negative numbers -> hence the uint to int conversion is not preserving the original value and therefore the hash != textureHash would always return true and each texture in the atlas was therefore considered dirty in the cache and updated every frame. 

Changing so hash is int as well to keep in line with C# preferring ints and to avoid the need of casting giving that the hash function actually returns an int. 

- The next check 

`            else if (m_IsGPUTextureUpToDate.TryGetValue(key, out var value))
                return value == 0 || (needMips && value == 1);
`

would also always return true (which means that an update is needed), the reason for that is that the validity was indexed by Key here which is the result of the GetTextureHash() or some derivative of it, _however_ when marking a texture as up to date in that dictionary, it was done using `texture.GetInstanceID()` as key, which is always different than whatever comes from GetTextureHash() .

Given how maps and dictionaries work when marking for validity this way:
`            m_IsGPUTextureUpToDate[instanceId] = (mipAreValid) ? 2 : 1; `

An new entry with instanceId is created in the map, instead of updating the one we wanted. So if a valid texture had InstanceID() = X and a hash = Y, we would check a different record even though is the same texture. 

So this condition also made it always return true to the question "Needs Update" 

This is not all, because now with proper checks in place we introduce a regression behaviour, if the texture content changed before because the texture was always updated, the entry in the atlas would of course be correct. To have the same behaviour we now need to check the content of the image hash ( https://github.com/Unity-Technologies/Graphics/compare/HDRP/staging...HDRP/fix-several-issues-with-texture-atlas?expand=1#diff-933d73095e66f1cc1bc7b4a8ad8dd766R310 ) 

Finally, also since we were using the _hash_ as both key and value, we'd never be able to check previous results, so keeping that would mean adding a new texture entry instead of updating the old one when a change happened. The reason for this is that as soon as an old texture is changed, its hash changes too, so we cannot use as a key to check for previous value. 
To solve for this, now the key is always set to `GetInstanceID()` and the value of the maps is still the hash . 
 

___________________________

**For QA:** I tested basic cookie workflow (added new cookies, modified cookies, deleted lights with cookies and added new ones) and tested changing the content of a texture that is already used as cookie from an image editing program and verify that it updates correctly without inserting a new entry in the atlas (verified via the atlas view in debug window), however given the nature of the change, I'd feel more comfortable if we could give this a more thorough check than just mine.

### CODE REVIEW

**Comment by FrancescoC-unity** [comment]
Gonna kick yamato on this one
